### PR TITLE
Add grid placement tool

### PR DESCRIPTION
**Pull request by OliBomby**
Part of https://github.com/ppy/osu/pull/26303

Requires
- [x] https://github.com/ppy/osu/pull/26309
- [x] https://github.com/ppy/osu/pull/26310
- [x] https://github.com/ppy/osu/pull/26311
- [x] https://github.com/ppy/osu/pull/29986

Adds the 'Change grid' placement tool to the left tool box.

Mappers might want to align their grid with some hit objects they have already placed, so I also added a 'Grid from points' feature which allows the mapper to select two points on the playfield and then the grid aligns itself with those points. The origin is centered on the first point and then the spacing and rotation is determined by the difference between the two points. The two points will be on two neighbouring grid lines in the resulting grid.
This point selection has snapping to nearby objects so its easy to align it exactly with hit objects.
Also clicking with right click will end the point selection immediately so you can select just one point allowing you to change just the grid origin.

I repurposed the initial PR https://github.com/ppy/osu/pull/26303 to add the remaining keybinds for cycling grid type.

https://github.com/user-attachments/assets/487a7d26-c420-4e33-91e3-ecee6946f958

### CODE REVIEW

**Comment by bdach** [comment]
Feature seems useful *but* as usual I have UX concerns.

https://github.com/user-attachments/assets/e41f9593-e506-40db-b7ab-bf289c3d5443

There is no guidance on how to use this feature. There is a button with no tooltip. You click on it and nothing visually happens. You click on it five times and maybe move it over the playfield and notice the grid starts moving. Only at that point you can probably figure out that you need to click the playfield twice to have it work. I'd still say both points should be visually indicated somehow rather than just relying on the grid but maybe that's optional.

The main problem is no indication on what you're supposed to do with the feature at all.

Also there is no way to cancel a grid change if you decide you don't want to change the grid anymore. Esc doesn't work.

Also I don't get what the criteria for the grid spacing subdividing is. It appears that it will decrease when the cursor moved too much between clicks 1 & 2, but I have no idea why it's doing what it's doing in particular.

There's also whatever this is:

https://github.com/user-attachments/assets/44148c5f-a453-4c8a-8386-7e4965d17cfa

You can be in the "change grid" mode and still have things like object placement appear to work but they actually will not work and instead the grid change thing will take priority. At that point I'd start thinking if this shouldn't be a tool in the toolbox on the left.

**Comment by OliBomby** [comment]
I changed the 'grid from points' button into a placement tool like @bdach suggested. I think the UX is really nice now.

The diffstat became pretty big but most of it is from 1a81e12192b6b6d2c1c8d1b9256cb2242e8d66e3 which is mostly just refactoring name change. It should be pretty easy to review as that commit should not change any behaviours.

**Comment by bdach** [comment]
The UX seems good now but dear lord the diffstat.

At this point I'd want someone to cross-check whether this is the correct direction UX-wise (hi @peppy again) and then if it turns out it is, then I'd want the huge rename commit to be split out of this into a separate PR.

**Comment by peppy** [comment]
It's a bit weird as a placement tool but I think I could get used to it.

That said, wouldn't a click + drag feel more intuitive? Click and release will just move the origin, click and drag will set origin and change rotation? Then we can do away with the left-left-click and right-click nonsense.

**Review by peppy — COMMENTED** [review]
As proposed.

**Comment by OliBomby** [comment]
> That said, wouldn't a click + drag feel more intuitive? Click and release will just move the origin, click and drag will set origin and change rotation? Then we can do away with the left-left-click and right-click nonsense.

I prefer the UX without drag because not holding down the mouse button makes it easier to do some seeks in between clicking the two points which is useful if they are some ways away on the timeline.
Also it shouldn't be that unintuitive because it follows the same control scheme as other placement tools, like right-click is typically associated with ending the interaction. You can end the interaction on the first click to only set the grid origin.

**Comment by peppy** [comment]
> Also it shouldn't be that unintuitive because it follows the same control scheme as other placement tools, like right-click is typically associated with ending the interaction. You can end the interaction on the first click to only set the grid origin.

The thing is, the UX for other tools is pretty unintuitive and I'd probably be looking to change the default (like what we've done for slider dragging, and for other rulesets like mania long notes).

**Comment by minetoblend** [comment]
Personally I think that drag is a bit more intuitive here. However, oli raises a valid point here about being able to use the mouse for other things such as seeking in the timeline when using 2 clicks. This is something I frequently do when snapping hitobjects to each other so I can see it beeing very useful for placing grids too. Imo the benefit from that outweighs the UX improvement that click+drag would create.

I want to propose an alternative though, which I think would be the best of both worlds.
What if the placement blueprint generally uses click+drag, but also draws 2 draggable handles (similar to the ones on the selection box) for moving the grid origin & the point dictating rotation & scale. Using those handles wouldn't immediately complete the interaction, so you can make multiple adjustments & seek to different points on the timeline in-between.

**Comment by OliBomby** [comment]
> I want to propose an alternative though, which I think would be the best of both worlds. What if the placement blueprint generally uses click+drag, but also places 2 draggable handles (similar to the ones on the selection box) on the playfield for moving the grid origin & the point dictating rotation & scale. Using those handles wouldn't immediately complete the interaction, so you can make multiple adjustments & seek to different points on the timeline in-between.

It's not really the best of both worlds. The interaction would become very tedious for the case where you want to snap to 2 points far apart on the timeline:
- I drag from point A to somewhere near point B
- Interaction ends
- I start the interaction again
- Draggable handle appears in a different location far away from point B due to spacing subdivision
- Seek to point B on the timeline
- Drag the handle a huge distance to point B
- Manually end interaction

An alternative that I think better combines the best of both worlds:
- 2 left clicks for setting origin, spacing and rotation
- 2 left clicks in the same location for setting origin
- 1 right click for setting origin
- 1 drag for setting origin, spacing, and rotation

This way we'd have the intuitive click+drag interaction as well as the convenient 2 click interaction. The only thing we cant have is the single left click and release for setting origin.

**Comment by minetoblend** [comment]
> * Interaction ends
> * I start the interaction again

I don't think the interaction should end if you move one of the handles, it should only end immediately when using click+drag on the playfield.

**Comment by OliBomby** [comment]
> > * Interaction ends
> > * I start the interaction again
> 
> I don't think the interaction should end if you move one of the handles, it should only end automatically when using click+drag on the playfield.

That part is using the click+drag on the playfield, so the interaction ends.

**Comment by minetoblend** [comment]
I see. I wasn't really thinking of the case of someone using click+drag, and then re-adjusting afterwards.

**Comment by OliBomby** [comment]
I've placed the grid placement tool back in the right toolbox, because its conceptually closer to the rotation/scale tools and it doesn't need to be as prominent as the hitobject tools.

I changed the behaviour of the grid placement tool to combine both dragging and clicking:
- 2 left clicks for setting origin, spacing and rotation
- 2 left clicks in the same location for setting origin
- 1 right click for setting origin
- 1 drag for setting origin, spacing, and rotation

I didn't go with draggable handles because I can't figure out a reasonable place to put the second draggable handle in all cases. If the grid spacing is low then the handles will be too close too eachother, and if you drag the handles to make the grid spacing is big, the spacing subdivision makes it so the draggable handles wont be in the same place you left them.

**Comment by bdach** [comment]
> I've placed the grid placement tool back in the right toolbox, because its conceptually closer to the rotation/scale tools and it doesn't need to be as prominent as the hitobject tools

Not sure about this... Sure, in this state maybe having it a tool still addresses the previous concern about placements being bugged, and I could probably agree with the fact that it's not as important as the other tools, _and_ I don't even entirely hate the fact that choosing it deselects all of the "conventional" tools - _but_ the UX of the *actual button* is terrible again because there is again no conveyance of what to do after you click it, no visible state, nothing.

Maybe if that were to be improved I'd be ok with it.

**Comment by OliBomby** [comment]
I added a tooltip so it tells you what to do while you are hovering the button wondering what to do. From then on it should be pretty obvious how it works.

File: osu.Game/Graphics/UserInterfaceV2/RoundedButton.cs [hunk 1]
```diff
@@ -107,5 +108,7 @@ public bool MatchingFilter
         }
 
         public bool FilteringActive { get; set; }
+
+        public virtual LocalisableString TooltipText { get; set; }
```
Comment (bdach) [inline osu.Game/Graphics/UserInterfaceV2/RoundedButton.cs:112]: This entire change with making `TooltipText` virtual is quite dodgy because some derived classes override the getter but leave the setter intact. Previously the intent was that the tooltip in those classes was derived from internal data and as such is immutable outside the class. But now with this you can freely overwrite the `TooltipText` of a `ReadyButton` or `FavouriteButton` or whatever else which is improper design.

The easiest way out of this would probably be to drop the `virtual`, and update the places that override the getter so that instead of it they use the setter (which will require a few bindable subscriptions or something to provide the same effect).

File: osu.Game/Rulesets/Edit/PlacementBlueprint.cs [hunk 2]
```diff
@@ -71,6 +71,8 @@ public virtual void EndPlacement(bool commit)
             PlacementActive = PlacementState.Finished;
         }
 
+        public virtual SnapType SnapType => SnapType.All;
```
Comment (bdach) [inline osu.Game/Rulesets/Edit/PlacementBlueprint.cs:74]: Needs xmldoc

**Review by bdach — COMMENTED** [review]
I guess UX wise I'd be _fine_ with this in its current state, but I'd want @peppy's reading of it too to proceed with merging this in.

File: osu.Game/Graphics/UserInterfaceV2/RoundedButton.cs (hunk 1 above)
Comment (OliBomby) [reply osu.Game/Graphics/UserInterfaceV2/RoundedButton.cs:112]: This might be a good place to use the `new` keyword. By declaring a `public new LocalisableString TooltipText` in `ReadyButton` and `FavouriteButton` we can remove the setter entirely which seems more appropriate for these classes which want to manage their tooltip text internally.

Comment (bdach) [reply osu.Game/Graphics/UserInterfaceV2/RoundedButton.cs]: Eh. I don't like shadowing because it makes things tricky anyway. Simple non-virtual non-shadowed public `get; set;` is best.

Comment (OliBomby) [reply osu.Game/Graphics/UserInterfaceV2/RoundedButton.cs]: I would agree with you that shadowing makes things tricky in most cases, but in this case using shadowing seems like the simplest solution. If we just considered `ReadyButton` and `FavouriteButton` it would be quite a simple change because they use only bindables for their tooltip. However we have to consider the tooltip of `ReadyButton` actually gets overriden by `PlaylistsReadyButton` and `MultiplayerReadyButton` which each use several non-bindable variables in determining their tooltip text.
I'd have to add new bindables in several classes and it would become a quite complicated chain of bindable updates.

**Comment by peppy** [comment]
No idea about the tooltip changes I just wangs some shit to make it compile so I could comment on the UX.

- I don't think the right-click method needs to exist. You can just left-click twice to get the same behaviour right?
- I preferred this being on the left, as it clears the user's selection and behaves like a modal operation but now has no indication that it will do that. Can we move it back?

**Review by peppy — COMMENTED** [review]
Updated UX feedback.

**Comment by OliBomby** [comment]
> * I don't think the right-click method needs to exist. You can just left-click twice to get the same behaviour right?

Yes left-click twice has the same behaviour, but I prefer to use a single right-click instead. It doesn't hurt to have a little redundancy to make it a bit easier to use right?

> * I preferred this being on the left, as it clears the user's selection and behaves like a modal operation but now has no indication that it will do that. Can we move it back?

It already clears the user's selection, but if you prefer it being on the left I'll move it there.

> * It seems to maintain a spacing under 64 while the limit is 128, is there a reason for that?

There's a good reason for that. I made the spacing match the range of the grid spacing cycle hotkey which is [4,64), so you can always cycle the spacing value back to the original value you got from the grid placement tool. I think these are all reasonable spacing values, beyond that there are so few snappable points on the screen that it is barely of any utility. The limit on the spacing slider is intentionally bigger for the off-chance you want to use a big spacing, so you can set that manually. 

The grid placement tool and the spacing cycling hotkey are tools that set the spacing automatically so I think it makes sense to restrict them to a range of more reasonable values.

**Comment by OliBomby** [comment]
> Let me rephrase: because it clears the selection, I think it should be on the left. Users are understanding that their selection will be lost when changing tools here. A button on the right randomly nuking a selection is unexpected.

That makes sense. I've moved it back to the left toolbox.

**Comment by peppy** [comment]
I'd either call the toolbox item "Grid" or "Place grid". The first may fit better with the other tools? @bdach thoughts?

Also I think the FA icon looks nicer:

```diff
diff --git a/osu.Game.Rulesets.Osu/Edit/GridFromPointsTool.cs b/osu.Game.Rulesets.Osu/Edit/GridFromPointsTool.cs
index bc143886ce..11c97d78d9 100644
--- a/osu.Game.Rulesets.Osu/Edit/GridFromPointsTool.cs
+++ b/osu.Game.Rulesets.Osu/Edit/GridFromPointsTool.cs
@@ -3,7 +3,6 @@
 
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Sprites;
-using osu.Game.Graphics;
 using osu.Game.Rulesets.Edit;
 using osu.Game.Rulesets.Edit.Tools;
 using osu.Game.Rulesets.Osu.Edit.Blueprints;
@@ -13,7 +12,7 @@ namespace osu.Game.Rulesets.Osu.Edit
     public partial class GridFromPointsTool : CompositionTool
     {
         public GridFromPointsTool()
-            : base("Change grid")
+            : base("Place grid")
         {
             TooltipText = """
                           Left click to set the origin.
@@ -23,7 +22,7 @@ Right click to only set the origin.
                           """;
         }
 
-        public override Drawable CreateIcon() => new SpriteIcon { Icon = OsuIcon.EditorPlaceGrid };
+        public override Drawable CreateIcon() => new SpriteIcon { Icon = FontAwesome.Solid.DraftingCompass };
 
         public override PlacementBlueprint CreatePlacementBlueprint() => new GridPlacementBlueprint();
     }

```

**Review by peppy — COMMENTED** [review]
As above

**Comment by OliBomby** [comment]
> I'd either call the toolbox item "Grid" or "Place grid". The first may fit better with the other tools? @bdach thoughts?
> 
> Also I think the FA icon looks nicer:

I agree. "Grid" fits better in the theme of the other placement tools.

**Comment by bdach** [comment]
> I'd either call the toolbox item "Grid" or "Place grid". The first may fit better with the other tools? @bdach thoughts?

Either's fine, would weakly prefer first.

**Review by peppy — APPROVED** [review]
